adapters/adhese/adhese.go

```diff
@@ -0,0 +1,265 @@
+		}
+
+		if err := json.Unmarshal(response.Body, &originDataArray); err != nil {
+			return nil, []error{err, WrapServerError(fmt.Sprintf("Response %v could not be parsed to JERLICIA origin data.", string(response.Body)))}
+		}
+		bidResponse = convertAdheseBid(adheseBid, extArray[0], originDataArray[0])
+	} else {
+		bidResponse = convertAdheseOpenRtbBid(adheseBid)
+	}
+
+	price, err := strconv.ParseFloat(adheseBid.Extension.Prebid.Cpm.Amount, 64)
+	if err != nil {
+		return nil, []error{err, WrapServerError(fmt.Sprintf("Could not parse Price %v as float ", string(adheseBid.Extension.Prebid.Cpm.Amount)))}
+	}
+	width, err := strconv.ParseUint(adheseBid.Width, 10, 64)
+	if err != nil {
+		return nil, []error{err, WrapServerError(fmt.Sprintf("Could not parse Width %v as int ", string(adheseBid.Width)))}
+	}
+	height, err := strconv.ParseUint(adheseBid.Height, 10, 64)
+	if err != nil {
+		return nil, []error{err, WrapServerError(fmt.Sprintf("Could not parse Height %v as int ", string(adheseBid.Height)))}
+	}
+	bidResponse.Cur = adheseBid.Extension.Prebid.Cpm.Currency
+	if len(bidResponse.SeatBid) > 0 && len(bidResponse.SeatBid[0].Bid) > 0 {
+		bidResponse.SeatBid[0].Bid[0].Price = price
+		bidResponse.SeatBid[0].Bid[0].W = width
+		bidResponse.SeatBid[0].Bid[0].H = height
+	}
+
+	bidderResponse := adapters.NewBidderResponseWithBidsCapacity(5)
+
+	if len(bidResponse.SeatBid) == 0 {
+		return nil, []error{WrapServerError("Response resulted in an empty seatBid array.")}
+	}
+
+	var errs []error
+	for _, sb := range bidResponse.SeatBid {
+		for i := 0; i < len(sb.Bid); i++ {
+			bid := sb.Bid[i]
+			bidderResponse.Bids = append(bidderResponse.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: getBidType(bid.AdM),
+			})
+
+		}
+	}
+	return bidderResponse, errs
+}
+
+func convertAdheseBid(adheseBid AdheseBid, adheseExt AdheseExt, adheseOriginData AdheseOriginData) openrtb.BidResponse {
+	adheseExtJson, err := json.Marshal(adheseOriginData)
+	if err != nil {
+		glog.Error(fmt.Sprintf("Unable to parse adhese Origin Data as JSON due to %v", err))
+		adheseExtJson = make([]byte, 0)
+	}
+	return openrtb.BidResponse{
+		ID: adheseExt.Id,
+		SeatBid: []openrtb.SeatBid{{
+			Bid: []openrtb.Bid{{
+				DealID: adheseExt.OrderId,
+				CrID:   adheseExt.Id,
+				AdM:    getAdMarkup(adheseBid, adheseExt),
+				Ext:    adheseExtJson,
+			}},
+			Seat: "",
+		}},
+	}
+}
+
+func convertAdheseOpenRtbBid(adheseBid AdheseBid) openrtb.BidResponse {
+	var response openrtb.BidResponse = adheseBid.OriginData
+	if len(response.SeatBid) > 0 && len(response.SeatBid[0].Bid) > 0 {
+		response.SeatBid[0].Bid[0].AdM = adheseBid.Body
+	}
+	return response
+}
+
+func getAdMarkup(adheseBid AdheseBid, adheseExt AdheseExt) string {
+	if adheseExt.Ext == "js" {
+		if ContainsAny(adheseBid.Body, []string{"<script", "<div", "<html"}) {
+			return adheseBid.Body + "<img src='" + adheseExt.ImpressionCounter + "' style='height:1px; width:1px; margin: -1px -1px; display:none;'/>"
```

Comment:
How does the ImpressionCounter handle privacy regulation? Is it possible for the url to be empty?

Reply:
I'm not the expert at Adhese. We fully implement the GDPR protocols which handles sensitive user data/tracking. As far as I'm aware we can track a bid in order to properly bill it - that is the only use of this counter. If you require so I can move it up the chain here and get a better answer formulated. In any case, it could be empty so I'll add it to the condition. I'm guessing that's the main reason to ask?

Reply:
Yes, exactly.